File: 2-introduction-to-deep-agents/typescript/m1/m1.6_homework_filled.ts
// typescript/m1/m1.6_homework_filled.ts
/**
 * Reference copy of m1.6_homework.ts with TODOs 1 and 2 filled in so you
 * can run it end to end and see what "done" looks like. This is just one
 * possible answer, so yours might be different. Explore!
 */

import { MultiServerMCPClient } from "@langchain/mcp-adapters";
import { createDeepAgent } from "deepagents";

import { model } from "../models.js";

// TODO 1 filled in: a different public MCP server than the lab's,
// X Docs, which searches X's public API documentation.
async function buildTools() {
  const client = new MultiServerMCPClient({
    "x-docs": {
      transport: "http",
      url: "https://docs.x.com/mcp",
    },
  });

  const tools = await client.getTools();

  const ALLOWED = new Set(["search_x"]);
  return { client, tools: tools.filter((t) => ALLOWED.has(t.name)) };
}

// TODO 2 filled in
const QUESTION =
  "Use the X Docs tool to search the X API documentation: what HTTP " +
  "status code and error code does the API return when a rate limit " +
  "is exceeded?";

const { client, tools } = await buildTools();

try {
  const agent = createDeepAgent({ model, tools });

  const result = await agent.invoke({
    messages: [{ role: "user", content: QUESTION }],
  });

  console.log(result.messages.at(-1)?.content);
} finally {
  await client.close();
}

File: 2-introduction-to-deep-agents/python/m1/m1.6_homework_filled.py
# python/m1/m1.6_homework_filled.py
"""Reference copy of m1.6_homework.py with TODOs 1 and 2 filled in so you
can run it end to end and see what "done" looks like. This is just one
possible answer, so yours might be different. Explore!"""

import asyncio
import warnings

warnings.filterwarnings("ignore", category=DeprecationWarning)

from deepagents import create_deep_agent
from langchain_mcp_adapters.client import MultiServerMCPClient

from models import model


# TODO 1 filled in: a different public MCP server than the lab's,
# DeepWiki, which answers questions about any public GitHub repo.
async def build_tools():
    client = MultiServerMCPClient({
        "deepwiki": {
            "transport": "http",
            "url": "https://mcp.deepwiki.com/mcp",
        }
    })
    tools = await client.get_tools()

    ALLOWED = {"ask_question"}
    return [t for t in tools if t.name in ALLOWED]


# TODO 2 filled in
QUESTION = (
    "Use the DeepWiki tool to ask the langchain-ai/deepagents GitHub repo: "
    "what filesystem backends does deepagents support?"
)


async def main():
    tools = await build_tools()
    agent = create_deep_agent(model=model, tools=tools)
    result = await agent.ainvoke({"messages": [{"role": "user", "content": QUESTION}]})
    print(result["messages"][-1].content)


asyncio.run(main())
